### ISSUE TITLE
Request New Language

### ISSUE THREAD

**Issue by jag3773** [unfoldingWord-dev/ts-android#807]
For some time we have discussed what users should do when they can't find their language in tS.

The kernel of the question is, should we allow them to request a new language in tS?  More verbosely, how do we prevent unnecessary creations becuase the user couldn't find the code that already exists for their language?  How do we prevent duplicates?

Potentially, we could create some "other" languages that people could choose.

Potentially, we could create a form similar to this:
* What is the name of your language?
* What are other names for your language?
* What are the names of similar languages?
* What area is this spoken in?
* Is there a language code for your language?

Those questions would allow us to:
1. Prompt them to select from a list of potential matches, or
2. Create a new private use code that would have the answers to all those questions attached to it.  When it makes it's way up to our servers we could have a surveyor review the information and add it to tD if necessary, or categorize it under an existing language code.

This is a terribly complicated issue, not so much on the technical side as it is on the survey side.  Open to any other ideas of how we could accomplish this.

**Comment by neutrinog** [comment]
@jag3773 any new regarding how we want to implement this?

**Comment by jag3773** [comment]
More details in https://github.com/unfoldingWord-dev/translationDatabaseWeb/issues/248#issuecomment-159047588.

**Comment by neutrinog** [comment]
This will eventually hit the TD api but for now we will store the info in the repo

**Comment by PhotoNomad0** [comment]
@jag3773 - you said you had some updated questions for this form?

**Comment by PhotoNomad0** [comment]
@jag3773 , @sampeckham - I found this questionnaire.  Is this what we should ask?

Questions for potential new languages
February 2016
1.	What do you call your language?
1.1. Does that have a special meaning?
1.2. Do you have other names for your language?

2.	Do other people have a different name for your language?
2.1. What do they call it?
2.2. Who calls it that?
2.3. Does that have a special meaning?

3.	Where else do people speak just the same way as you do?

4.	Where do people speak just a little bit differently?
4.1. Have you yourself gone to these places?
4.2. Do people from there also come here?
4.3. Do you have a name for this other dialect in your own language?

5.	Where do people speak very differently?
5.1. How much of their speech do you understand? (few words, main ideas, most, all)
5.2. Have you yourself gone to these places?
5.3. Do people from there come here?
5.4. Do you have a name for this other dialect in your own language?

6.	Where do people speak your own language the most purely?
6.1. Why do you say that is the most pure?
6.2. Have you yourself gone to these places?
6.3. Do people from there come here?

7.	Where is your language spoken badly?
7.1. Why do you say that it is spoken badly?
7.2. Have you yourself gone to these places?
7.3. Have people from there come here?

8.	What do you call the language people speak in [large place, 1-2  day’s march distance]?
8.1. Do people understand you when you speak your language there?
8.2. If a man from there come here, can everybody understand his speech?
8.3. Even the women and children?

9.	What other places do you travel to?
10.	Do many people travel outside from here?

**Comment by PhotoNomad0** [comment]
@neutrinog , @sampeckham  - since this is an interview, should this be laid out more like the publish activity with related questions in each section?

**Comment by jag3773** [comment]
Attached
[Short Language Survey questions.docx](https://github.com/unfoldingWord-dev/ts-android/files/142935/Short.Language.Survey.questions.docx)
 is the full list.

**Comment by PhotoNomad0** [comment]
I think they are going to need a nap after running that gauntlet of questions! ;)

**Comment by neutrinog** [comment]
@PhotoNomad0 yes we probably want to place the questionnaire in it's own activity and break it up into steps (fragments) like we do with publishing.

**Comment by neutrinog** [comment]
also I don't think we need to display at which step they are at in the questionnaire.

**Comment by PhotoNomad0** [comment]
@jag3773, @neutrinog  - I think one question needs to be rewritten. "2.3. Does that have a special meaning?" may lead to just a yes/no answer.  Maybe better to have something such as "2.3. Does that have a special meaning?  What is that meaning?" so that they would enter more info if it does have a special meaning.

**Comment by neutrinog** [comment]
The requirements issue for this is https://github.com/unfoldingWord-dev/ts-requirements/issues/20

**Comment by neutrinog** [comment]
@PhotoNomad0 there is a new-still-in-progress spec for this at https://github.com/unfoldingWord-dev/ts-requirements/wiki/New-Target-Language-Code. Still in progress but I think we can begin working on the language code generating part. And maybe half of recording questionnaire results.

**Comment by PhotoNomad0** [comment]
@jag3773 @neutrinog @RRohrssen - About the requirements doc "https://github.com/unfoldingWord-dev/ts-requirements/wiki/New-Target-Language-Code":

* Is the questionnaire API intended to be English only?  There doesn't seem to be a way to specify language for the questions.
* there also should be a list of the questions that require answers.
* just to verify - it looks like under meta/order that the questions id are grouped as an array of arrays.  Is this like a questions per page arrangement?
* the web api is not functional yet: http://td.unfoldingword.org/api/new-language-questions/
* submitting questionnaires seems to be the big hole yet in the spec.

**Comment by neutrinog** [comment]
@PhotoNomad0 

The api will be updated (eventually) to support multiple languages. In answer to all of your other questions, the api spec is not finished yet. So no api, no final draft, and no endpoint to receive the questionnaires. Stay tuned!

**Comment by RRohrssen** [comment]
Waiting for updated tD API. Expected within 1 - 2 weeks.

**Comment by neutrinog** [comment]
the new api is at http://td.unfoldingword.org/api/questionnaire

**Comment by RRohrssen** [comment]
The new TD API will send a list of questions. We are still waiting for the capability to send answers to those questions back to TD.

**Comment by PhotoNomad0** [comment]
@neutrinog - has this been added to the library for offline use?  It has been so long since we talked about this that I am not sure of the final decisions we made.

**Comment by neutrinog** [comment]
Not not yet. I actually forgot I needed to do that.

**Comment by PhotoNomad0** [comment]
@neutrinog - no problem, there is much more that I have to finish up.  At least now I have the pattern for the json.

**Comment by PhotoNomad0** [comment]
@neutrinog - under the **"Recording Questionnaires"**
...
  "requester": "", -- name of the translator who completed the questionnaire

For requester should I use this? : **_AppContext.getProfile().gogsUser.fullName_**

**Comment by PhotoNomad0** [comment]
@neutrinog - maybe we should use gogs username instead.

**Comment by neutrinog** [comment]
No use the profile full name. The profile already manages getting the
correct name if a gogs account is attached or not.

**Comment by PhotoNomad0** [comment]
Need New language API to be added to app.

**Comment by neutrinog** [comment]
After the questionnaire has been completed we need to display a confirmation to the user so they know what happened and so they can see their code (they may need to share it with someone else).

```
A temporary language code has been created for your translation.
my-language-code

[ok]
```

**Comment by neutrinog** [comment]
this is currently only blocked by https://github.com/unfoldingWord-dev/translationDatabaseWeb/issues/603

**Comment by neutrinog** [comment]
@jag3773 , @klappy  after a temporary language code is created on a device it will become available when creating new target translations in the future. Should we make a visual distinction between actual languages and temporary languages when viewed in a list?

**Comment by neutrinog** [comment]
@jag3773 @klappy **bump**

**Comment by neutrinog** [comment]
A lot of this has been completed. There are however a number of items that need to be completed with regard to management of data:

- [ ] Need to update the `new_language.json` file with the correct `submitted_at` time in a target translation that is imported if the language has already been submitted.
- [ ] Need to import `new_language.json` from a target translation that is imported if the langauge has not yet been submitted. The language requested will be stored in the data path and the temp language will be stored in the index.
- [ ] Need to migrate temp language codes to the approved ones whenever the list of approved temp languages has been updated. This may occur when updating the library.

**Comment by neutrinog** [comment]
The language changing part of this is related to https://github.com/unfoldingWord-dev/ts-requirements/issues/33
